File: hms-api/src/main/java/com/pms/masters/service/IpBillingCategoryService.java
package com.pms.masters.service;

import com.pms.common.EntityNotFoundException;
import com.pms.masters.dto.IpBillingCategoryDto;
import com.pms.masters.entity.IpBillingCategory;
import com.pms.masters.entity.RevenueBucket;
import com.pms.masters.repository.IpBillingCategoryRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class IpBillingCategoryService {

    private final IpBillingCategoryRepository repository;

    public IpBillingCategoryService(IpBillingCategoryRepository repository) {
        this.repository = repository;
    }

    public List<IpBillingCategoryDto> findActive() {
        return repository.findByActiveTrueOrderByNameAsc().stream().map(this::toDto).toList();
    }

    public List<IpBillingCategoryDto> findInactive() {
        return repository.findByActiveFalseOrderByUpdatedAtDesc().stream().map(this::toDto).toList();
    }

    @Transactional
    public IpBillingCategoryDto create(IpBillingCategoryDto dto) {
        if (repository.existsByNameIgnoreCase(dto.name())) {
            throw new IllegalArgumentException("IP Billing Category already exists: " + dto.name());
        }
        IpBillingCategory category = new IpBillingCategory();
        category.setName(dto.name());
        category.setActive(true);
        category.setRevenueBucket(dto.revenueBucket() != null ? dto.revenueBucket() : RevenueBucket.OTHER);
        return toDto(repository.save(category));
    }

    // Deliberately does not touch revenueBucket - the frontend's rename flow
    // (shared MasterCrudComponent) only ever sends a name, so accepting
    // dto.revenueBucket() here would silently reset an admin's earlier
    // CEO/MD Dashboard tagging back to OTHER on every plain rename. Setting
    // the bucket goes through updateRevenueBucket() instead.
    @Transactional
    public IpBillingCategoryDto update(Long id, IpBillingCategoryDto dto) {
        IpBillingCategory category = getOrThrow(id);
        category.setName(dto.name());
        return toDto(repository.save(category));
    }

    @Transactional
    public IpBillingCategoryDto updateRevenueBucket(Long id, RevenueBucket revenueBucket) {
        IpBillingCategory category = getOrThrow(id);
        category.setRevenueBucket(revenueBucket);
        return toDto(repository.save(category));
    }

    @Transactional
    public void deactivate(Long id) {
        IpBillingCategory category = getOrThrow(id);
        category.setActive(false);
        repository.save(category);
    }

    @Transactional
    public void restore(Long id) {
        IpBillingCategory category = getOrThrow(id);
        category.setActive(true);
        repository.save(category);
    }

    private IpBillingCategory getOrThrow(Long id) {
        return repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("IP Billing Category not found: " + id));
    }

    private IpBillingCategoryDto toDto(IpBillingCategory category) {
        return new IpBillingCategoryDto(category.getId(), category.getName(), category.isActive(), category.getRevenueBucket());
    }
}

File: hms-web/src/app/features/masters-admin/ip-billing-categories/ip-billing-category.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { RevenueBucket } from '../../../shared/revenue-bucket.model';
import { IpBillingCategory } from './ip-billing-category.model';

@Injectable({ providedIn: 'root' })
export class IpBillingCategoryService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/masters/ip-billing-categories`;

  list(): Observable<IpBillingCategory[]> {
    return this.http.get<IpBillingCategory[]>(this.baseUrl);
  }

  listInactive(): Observable<IpBillingCategory[]> {
    return this.http.get<IpBillingCategory[]>(`${this.baseUrl}/inactive`);
  }

  create(name: string): Observable<IpBillingCategory> {
    return this.http.post<IpBillingCategory>(this.baseUrl, { name });
  }

  update(id: number, name: string): Observable<IpBillingCategory> {
    return this.http.put<IpBillingCategory>(`${this.baseUrl}/${id}`, { name });
  }

  updateRevenueBucket(id: number, revenueBucket: RevenueBucket): Observable<IpBillingCategory> {
    return this.http.patch<IpBillingCategory>(`${this.baseUrl}/${id}/revenue-bucket`, { revenueBucket });
  }

  deactivate(id: number): Observable<void> {
    return this.http.patch<void>(`${this.baseUrl}/${id}/deactivate`, {});
  }

  restore(id: number): Observable<void> {
    return this.http.patch<void>(`${this.baseUrl}/${id}/restore`, {});
  }
}
